src/psyclone/psyir/backend/sympy_writer.py

```diff
@@ -115,86 +132,212 @@ def __init__(self, type_map=None):
+                        if args[i] == args[i+1] and args[i+2] == "1":
+                            # a(i,i,1) --> a(i)
+                            new_args.append(args[i])
+                        elif args[i] == lower_b and args[i+1] == upper_b and \
+                                args[i+2] == "1":
+                            # a(lower_b, upper_b, 1) --> a(:)
+                            new_args.append(":")
+                        else:
+                            if args[i+2] == "1":
+                                # a(i,j,1) --> a(i:j)
+                                new_args.append(f"{args[i]}:{args[i+1]}")
+                            else:
+                                # a(i,j,k) --> a(i:j:k)
+                                new_args.append(f"{args[i]}:{args[i+1]}:"
+                                                f"{args[i+2]}")
+                    return f"{name}({','.join(new_args)})"
+
+                self._sympy_type_map[name] = \
+                    type(name, (Function, ),
+                         {"_sympystr": print_fortran_array})
+
+        # Now all symbols have been added to the symbol table, create
+        # unique names for the lower- and upper-bounds using special tags:
+        self._lower_bound = \
+            self._symbol_table.new_symbol("sympy_lower",
+                                          tag="sympy!lower_bound").name
+        self._upper_bound = \
+            self._symbol_table.new_symbol("sympy_upper",
+                                          tag="sympy!upper_bound").name
+
+    # -------------------------------------------------------------------------
+    @property
+    def lower_bound_name(self):
+        ''':returns: the name to be used for an unspecified lower bound.
+        :rtype: str
 
-        return sympy_type_map
+        '''
+        return self._lower_bound
+
+    # -------------------------------------------------------------------------
+    @property
+    def upper_bound_name(self):
+        ''':returns: the name to be used for an unspecified upper bound.
+        :rtype: str
 
-    @staticmethod
-    def get_sympy_expressions_and_symbol_map(list_of_expressions):
         '''
-        This function takes a list of PSyIR expressions, and converts
-        them all into Sympy expressions using the SymPy parser.
-        It takes care of all Fortran specific conversion required (e.g.
-        constants with kind specification, ...), including the renaming of
-        member accesses, as described in
-        https://psyclone-dev.readthedocs.io/en/latest/sympy.html#sympy
-        It also returns the symbol map, i.e. the mapping of Fortran symbol
-        names to SymPy Symbols.
+        return self._upper_bound
+
+    # -------------------------------------------------------------------------
+    @property
+    def type_map(self):
+        ''':returns: the mapping of names to SymPy symbols or functions.
+        :rtype: Dict[str, :py:class:`sympy.core.symbol.Symbol`]
```

Comment:
I'm having a hard time following what's going on but at L265 you appear to fill entries in this dictionary with Functions, not Symbols?

Reply:
I've updated the doc string.